### PR TITLE
Add support for custom roles in website, beyond "Team leader"

### PR DESCRIPTION
**Pull request by dtolnay**
I would like to render "Editor" below @JoelMarcey's entry in https://www.rust-lang.org/governance/teams/lang#spec, in the same location that others have "Team leader".

Source: https://blog.rust-lang.org/inside-rust/2023/11/15/spec-vision.html#specification-team

Output of `cargo run -- dump-website` for translations:

```console
...
governance-team-wg-triage-name = Triage working group
governance-team-wg-triage-description = Triaging repositories under the rust-lang organisation

governance-team-wg-wasm-name = WebAssembly (WASM) working group
governance-team-wg-wasm-description = Improving on the end-to-end experience of embedding Rust code in JS libraries and apps via WebAssembly

governance-role-spec-editor = Editor
```

Output of `cargo run -- static-api` under v1/teams/spec.json:

```json
{
  "name": "spec",
  "kind": "team",
  "subteam_of": "lang",
  "members": [
    {
      "name": "Mara Bos",
      "github": "m-ou-se",
      "github_id": 783247,
      "is_lead": true
    },
    {
      "name": "Felix Klock",
      "github": "pnkfelix",
      "github_id": 173127,
      "is_lead": true
    },
    {
      "name": "Eric Huss",
      "github": "ehuss",
      "github_id": 43198,
      "is_lead": false
    },
    {
      "name": "Joel Marcey",
      "github": "JoelMarcey",
      "github_id": 3757713,
      "is_lead": false,
      "roles": [
        "spec-editor"
      ]
    }
  ],
```

### CODE REVIEW

**Comment by dtolnay** [comment]
Another possible application: team affiliation in https://www.rust-lang.org/governance/teams/leadership-council#leadership-council. "Lang team rep"

File: src/static_api.rs [hunk 1]
```diff
@@ -121,6 +135,9 @@ impl<'a> Generator<'a> {
                         github: (*github_name).into(),
                         github_id: person.github_id(),
                         is_lead: leads.contains(github_name),
+                        roles: website_roles
+                            .get(*github_name)
+                            .map_or_else(Vec::new, Vec::clone),
```
Comment (rylev) [inline src/static_api.rs:140]: Nit: this might be slightly more clear:

```suggestion
                            .cloned().unwrap_or_default()
```

File: src/validate.rs [hunk 2]
```diff
@@ -774,6 +776,64 @@ fn validate_repos(data: &Data, errors: &mut Vec<String>) {
...
+/// (because the role id becomes the Fluent id used for translation).
+fn validate_website_roles(data: &Data, errors: &mut Vec<String>) {
+    let mut role_descriptions = HashMap::new();
+
+    wrapper(
+        data.teams().chain(data.archived_teams()),
+        errors,
+        |team, errors| {
+            let team_name = team.name();
+            if let Some(website_data) = team.website_data() {
```
Comment (rylev) [inline src/validate.rs:790]: Nit: might help if we try to reduce the rightward drift a bit:

```suggestion
           let Some(website_data) = team.website_data() else { return Ok(()); };
```

**Review by rylev — APPROVED** [review]
Thanks for working on this! Overall it's a great change, and I'd be happy to merge except for one thing.

The current schema requires duplicating the member's name both in the team's `members` array as well as part of role's `members` array. Would it make sense to see if we can change the team `members` array schema to allow for specifying roles inline to the member declaration?

Perhaps something like:

```toml
members = [
  "rylev",
  { name =  "dtolnay", role = "foo" }
]
```

According to the [toml spec](https://github.com/toml-lang/toml/blob/1.0.0/toml.md#array) this should be fine, but I think maybe some toml parsers struggle with this syntax?

File: src/static_api.rs (hunk 1 above)
Comment (dtolnay) [reply src/static_api.rs]: Fixed.

File: src/validate.rs (hunk 2 above)
Comment (dtolnay) [reply src/validate.rs]: Done. Good call.

**Comment by dtolnay** [comment]
Updated to set roles inline in the team's existing `members` array, as you suggested.

File: src/schema.rs [hunk 3]
```diff
@@ -521,6 +528,34 @@ pub(crate) struct TeamPeople {
     pub include_all_alumni: bool,
 }
 
+#[derive(serde::Deserialize, Clone, Debug)]
+#[serde(remote = "Self")]
+pub(crate) struct TeamMember {
+    #[serde(rename = "name")]
+    pub github: String,
```
Comment (rylev) [inline src/schema.rs:535]: I know I suggested calling this "name", but I do wonder if using "github" as the key would be better. This is what it's called in the people toml config whereas "name" is used for the person's preferred name.

File: docs/toml-schema.md [hunk 4]
```diff
@@ -133,6 +139,14 @@ zulip-stream = "t-lang"
 # Default is 0.
 weight = -100
 
+# Customized roles held by a subset of the team's members, beyond "Team leader"
+# which is rendered automatically for members of the `leads` array.
+[[website.roles]]
```
Comment (rylev) [inline docs/toml-schema.md:144]: Namespacing these roles under website feels a bit strange. I know the reason for this PR is to display this information on the website, but is this information only relevant to the website? Perhaps this should just be a top level `roles` array?

File: docs/toml-schema.md [hunk 5]
```diff
@@ -49,6 +49,12 @@ members = [
     "rfcbot",
     "craterbot",
     "rust-timer",
+    # Any subset of members may hold custom roles, beyond "Team leader" which is
+    # controlled by the `leads` array above. See [[website.roles]]. Members with
+    # roles are written using an inline table as follows. A simple string member
+    # like "bors" is equivalent to {name = "bors", roles = []}.
```
Comment (rylev) [inline docs/toml-schema.md:55]: Can we mention that the roles listed with each member must also be present in the roles array?

**Review by rylev — COMMENTED** [review]
Awesome! I think we can merge as is as the static API looks good (and that's the only thing that is somewhat tricky to change after the fact). However, I'd like to get your opinion on my comments on the schema.

File: docs/toml-schema.md (hunk 4 above)
Comment (dtolnay) [reply docs/toml-schema.md]: Fixed.

File: docs/toml-schema.md (hunk 5 above)
Comment (dtolnay) [reply docs/toml-schema.md]: Done.

File: src/schema.rs [hunk 6]
```diff
@@ -601,6 +641,13 @@ impl WebsiteData {
     }
 }
 
+#[derive(serde_derive::Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct MemberRole {
```
Comment (dtolnay) [inline src/schema.rs:646]: `MemberRole` in contrast with `DiscordRole`. Not sure if this is a great name. `Role` did not seem adequate.

https://github.com/rust-lang/team/blob/57e50ecf446550dc0ef48f8e48bcc3b8abada10e/src/schema.rs#L465-L468

**Review by rylev — APPROVED** [review]
I think this is a great start! Let's merge this and we can adjust from here.

@dtolnay seems like there might be conflicts that need to be resolved first though.

**Comment by dtolnay** [comment]
Rebased to fix superficial conflict with #1164 in src/validate.rs.

**Review by rylev — APPROVED** [review]
